source/common/common/mem_block_builder.h

```diff
@@ -0,0 +1,136 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+
+// Manages a block of raw memory for objects of type T. T is generally expected
+// to be a POD, where it makes sense to memcpy over it. This class carries extra
+// member variables for tracking size, and a write-pointer to support safe
+// appends.
+//
+// MemBlockBuilder is used to safely write blocks of data into a memory
+// buffer. Due to two extra member variables, it is not optimal for storing in
+// data structures. The intended usage model is to release the raw assembled
+// memory block from the MemBlockBuilder for efficient storage.
+//
+// The goal for this class is to provide a usage model to replace direct usage
+// of memcpy with a pattern that is easy to validate for correctness by
+// inspection, asserts, and fuzzing, but when compiled for optimization is
+// roughly as efficient as raw memcpy.
+template <class T> class MemBlockBuilder {
+public:
+  // Constructs a MemBlockBuilder allowing for 'capacity' instances of T.
+  explicit MemBlockBuilder(uint64_t capacity)
+      : data_(std::make_unique<T[]>(capacity)), capacity_(capacity), capacity_remaining_(capacity),
+        write_ptr_(data_.get()) {}
+  MemBlockBuilder() : capacity_(0), capacity_remaining_(0), write_ptr_(nullptr) {}
+
+  /**
+   * Populates (or repopulates) the MemBlockBuilder to make it the specified
+   * capacity. This does not have resize semantics; when populate() is called any
+   * previous contents are erased.
+   *
+   * @param capacity The number of memory elements to allocate.
+   */
+  void populate(uint64_t capacity) {
+    data_ = std::make_unique<T[]>(capacity);
+    capacity_ = capacity;
+    capacity_remaining_ = capacity;
+    write_ptr_ = data_.get();
+  }
+
+  /**
+   * @return the capacity.
+   */
+  uint64_t capacity() const { return capacity_; }
+
+  /**
+   * Appends a single object of type T, moving an internal write-pointer
+   * forward. Asserts that there is room to write the object when compiled
+   * for debug.
+   *
+   * @param object the object to append.
+   */
+  void appendOne(T object) {
+    RELEASE_ASSERT(capacity_remaining_ >= 1, "insufficient capacity");
+    ASSERT(write_ptr_ < (data_.get() + capacity_));
+    *write_ptr_++ = object;
+    --capacity_remaining_;
+  }
+
+  /**
+   * Appends raw data, moving an internal write-pointer forward. Asserts
+   * that there is room to write the block when compiled for debug. It is
+   * the caller's responsibility to ensure that the input data is valid.
+   *
+   * @param data The block of objects to insert.
+   * @param size The number of elements in the block.
+   */
+  void appendData(const T* data, uint64_t size) {
```

Comment:
How about using `absl::span` here? Seems more flexible and a bit more type checked for many uses?

Reply:
I'm glad you mentioned this because I did not know about `absl::Span<>`. Pretty cool; I'll keep that one in mind. But I don't see how that data structure helps in this case, as the functionality it automates isn't really what this class does, which is provide a builder-like interface for filling in an owned piece of data.

Reply:
You are passing `const T* data, uint64_t size` which can be expressed in a span, just like a string_view. It seemed a more flexible interface that does the same thing? nbd either way.

Reply:
I see...ok that definitely improves the interface. Thanks!

I also did -- in a separate commit -- a change to use a span in the rep as well. That reduces the number of lines of code in this impl. Note than in absl::Span::remove_prefix() there is an assert() already, so I removed the SECURITY_ASSERT for now, and key off that in the death-tests.

I feel like the change in AppendOne from `*write_ptr_++ = ch` to `write_span_[0] = ch; write_span_.remove_prefix(1);` is a little less intuitive and might not compile to something quite as fast. WDYT? I can always roll back https://github.com/envoyproxy/envoy/pull/9306/commits/e0d6ee1f42a84d10cc49f6e9b496cf66eb6da22c